Use futures sub-crates explicitly

Byron opened the pull request:
This bring initial compile times (using `cargo check`) down from ~22s to ~14s, while reducing the amount of dependencies from 25 to 23.

Please let me know if you would like to see any changes.

**edit**: These amazing build time improvements cannot be reproduced anymore. Probably just a local fluke :/.

Byron committed:
Use futures sub-crates explicitly

taiki-e commented on Cargo.toml:
```diff
@@ -13,7 +13,10 @@ categories = ["asynchronous", "concurrency"]
-futures = { version = "0.3.4", default-features = false, features = ["std"] }
+futures-util = { version = "0.3.5", default-features = false, features = ["std"] }
+futures-channel = { version = "0.3.5", default-features = false, features = ["std"] }
+futures-core = { version = "0.3.5", default-features = false, features = ["std"] }
```
futures-util exports all items defined in futures-core, so we don't need to depend on futures-core directly.

Byron committed:
Don't depend on futures-core directy if futures-util can be used instead

taiki-e commented on Cargo.toml:
```diff
@@ -13,7 +13,10 @@ categories = ["asynchronous", "concurrency"]
 
 [dependencies]
-futures = { version = "0.3.4", default-features = false, features = ["std"] }
+futures-util = { version = "0.3.5", default-features = false, features = ["std"] }
```
```suggestion
futures-util = { version = "0.3.5", default-features = false, features = ["std", "sink"] }
```

In futures-util, [sink is not enabled by default](https://github.com/rust-lang/futures-rs/blob/master/futures-util/Cargo.toml#L22).

([CI passes because of the feature flag of the dev-dependencies leaks to normal dependencies.](https://github.com/rust-lang/cargo/issues/4866))

Byron committed:
futures-util needs the 'sink' feature enabled

From the review:
> In futures-util, sink is not enabled by default.
> (CI passes because of the feature flag of the dev-dependencies leaks to normal dependencies.)

Byron commented:
@taiki-e  Thanks for the review! Do you see a way to make CI fail? Or asked differently: How did you detect the missing 'sink' feature? Maybe CI can be tuned to do the same.

taiki-e commented:
> How did you detect the missing 'sink' feature? Maybe CI can be tuned to do the same.

By running `cargo check -Z features=dev_dep`, that can be detected (require nightly).

Byron committed:
Run `cargo check` without leaking dev-dependencies (nighly only)

That way, missing dependencies won't be masked by them being pulled in
by dev-dependencies.

Byron committed:
Prevent `cargo check` to be run twice

Not doing so actually prevents the check we want to work correctly.

Byron committed:
Fix workflow yaml to allow build & test steps to run as expected

stjepang commented on .github/workflows/build-and-test.yaml:
```diff
@@ -33,11 +33,19 @@ jobs:
             override: true
 
       - name: Run cargo check
+        if: "!startsWith(matrix.rust, 'nightly')"
```
It would be good to check everything (benches, bins, examples, tests) on nightly anyways.
```suggestion
```

Byron replied:
It's odd, apparently I don't fully understand the `-Z features=dev_dep`. Let me try again to achieve something like this:

> By running cargo check -Z features=dev_dep, that can be detected (require nightly).

Maybe it's enough to literally run it like that, without additional flags, in addition to what's always run.
Commit is coming up.

Byron committed:
Run a bare cargo check on nighly with a flag to help seeing missing features

Should implement what's suggested in this comment:

https://github.com/stjepang/blocking/pull/2#issuecomment-629777534

Byron commented:
Oh, interesting, now it does indeed fail the CI. Do you have suggestions on how to fix it? If not, I will install nightly and see what I can do.

taiki-e commented:
Oh, futures-channel also needs to enable`sink` feature.

Byron committed:
Add 'sink' feature to futures-channel to fix build
